### PR TITLE
Feature issue 636

### PR DESCRIPTION
**Pull request by romanbalayan**
###  Summary

Add implementation of proxy support on IncomingWebhook using custom agent. Also added example code for usage with either https-proxy-agent or tunnel. Implements feature request for issue #636 

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackapi/node-slack-sdk/blob/master/.github/contributing.md) and have done my best effort to follow them.
* [X] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).

### CODE REVIEW

**Comment by aoberoi** [comment]
@romanbalayan thanks so much for your contribution! i think this is great, as long as we don't end up repeating too much code if we can avoid it. also, i want to point out that there's currently an issue with the way proxies are handled that is being discussed in #642. depending on how fast that issue is addressed, the changes might overlap with the lines of code you have modified.

File: src/IncomingWebhook.ts [hunk 1]
```diff
@@ -152,3 +167,27 @@ function httpErrorWithOriginal(original: Error): IncomingWebhookHTTPError {
   error.original = original;
   return (error as IncomingWebhookHTTPError);
 }
+
+/**
+ * Detects whether an object is an Agent
+ */
+function isAgent(obj: any): obj is Agent {
```
[the remark this answers is not in the document]
Comment (romanbalayan) [reply src/IncomingWebhook.ts]: Got it. Was considering that, actually, but was not sure if I should change the code for WebClient

[the remark this answers is not in the document]
Comment (romanbalayan) [reply src/IncomingWebhook.ts]: Should I include the changes on this same pull request?

[the remark this answers is not in the document]
Comment (romanbalayan) [reply src/IncomingWebhook.ts]: Checked the change on #642 issue. The current implementation of isAgent is same as in the latest merge; checking if addRequest is a function.

File: examples/basic-incoming-webhook-custom-agent.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+ * Example for using an Incoming Webhook with custom agent
+ */
+
+const { IncomingWebhook } = require('../dist');
+
+// Get a URL by creating an app at <https://api.slack.com/apps?new_app=1>, and configuring an Incoming Webhook
+// It's always a good idea to keep sensitive data like the url outside your source code. Prefer environment variables.
+const url = process.env.SLACK_WEBHOOK_URL || '';
+
+if (!url) { console.log('You must specify a token to use this example'); process.exitCode = 1; return; }
```
Comment (aoberoi) [inline examples/basic-incoming-webhook-custom-agent.js]: Change “token” to “url” in the log message

**Comment by aoberoi** [comment]
hey @romanbalayan, thanks for the quick turnaround! i can see there's still an issue with a merge conflict on this PR and i had one comment about changing the string in a log. let me know when you're ready for me to review this PR again.

**Comment by romanbalayan** [comment]
Hey @aoberoi, caught me at the right time a while ago!

Also, I already fixed the log message on the latest commit. The merge request had a conflict due to the migration of `isAgent` and `agentForScheme` functions to util.ts. 

Will you be the one manually resolving the conflict on WebClient.ts?

**Comment by romanbalayan** [comment]
Hi @aoberoi,

Already resolved the merge conflict! Thanks!

**Comment by aoberoi** [comment]
thanks @romanbalayan, this is really great. i'm especially impressed with your attention to detail to make sure the functionality has good test coverage and persistence to keep this PR up to date through several changes on master.

giphy 1

**Pull request merged**